src/background/utils.js

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+window.SITE_CONTEXTS_STORAGE_KEY = "siteContexts";
+window.MOZILLA_VPN_SERVERS_KEY =  "mozillaVpnServers";
+
+
+/**
+ * Here you'll find utility functions for managing 
+ * site contexts and handling various tasks. 
+ */
+export const Utils = {
+
+  /**
+   * Retrieves the context for a specific origin from storage.
+   * @param {string} origin - The origin to retrieve the context for.
+   * @param {Map<string, object>} siteContexts - A map containing contexts for various origins.
+   * @returns {object|null} - The context for the origin, or null if not found.
+   */
+  getContextForOrigin(origin, siteContexts) {
+    return siteContexts.get(origin);
+  },
+
+  /**
+   * Retrieves the currently active tab.
+   * @returns {Promise<browser.tabs.Tab>} - The current active tab.
+   */
+  async getCurrentTab() {
+    let currentTab = await browser.tabs.query({currentWindow: true, active: true});
+    return currentTab[0];
+  },
+
+  /**
+   * Formats and retrieves the hostname from a given URL.
+   * @param {string} url - The URL to format.
+   * @returns {string} - The formatted hostname.
+   */
+  getFormattedHostname(url) {
+    // Handle sites being viewed in reader mode
+    // TODO... other prefixes(?)
+    const readerPrefix = 'about:reader?url=';
+    if (url.startsWith(readerPrefix)) {
+      const encodedUrl = url.slice(readerPrefix.length);
+      url = decodeURIComponent(encodedUrl);
+    }
+    
+    let hostname = this.getHostname(url);
+
+    // Use the entire URL if hostname is not valid (like about:debugging)
+    if (!hostname || hostname === "") {
+      hostname = url;
+    }
+    return hostname;
+  },
+
+  /**
+   * Extracts the hostname from a given URL.
+   * @param {string} url - The URL to extract the hostname from.
+   * @returns {string|null} - The extracted hostname or null if invalid.
+   */
+  getHostname(url) {
+    const urlObj = new URL(url);
+    if (urlObj && urlObj.hostname) {
```

Comment:
urlObj is unlikely to be null given it will throw an exception on a non valid url :)